### PR TITLE
fby35: gl: Support DIMM and PMIC reading via I3C

### PR DESCRIPTION
**Pull request by BonnieLo**
# Description:
Read DIMM related data (DIMM's temperature, PMIC power, and PMIC error) by I3C

# Motivation:
Currently, BIC read DIMM by using PECI command

# Test Plan:
1. Check DIMM temperature and PMIC power sensors - pass
2. Inject PMIC error and check SEL - pass

# Test Log:
1. Check DIMM temperature and PMIC power sensors root@bmc-oob:~# sensor-util slot1|grep -i DIMM
MB_DIMMA_TEMP_C              (0x5) :   34.50 C     | (ok)
MB_DIMMB_TEMP_C              (0x6) :   33.00 C     | (ok)
MB_DIMMC_TEMP_C              (0x7) :   33.25 C     | (ok)
MB_DIMMD_TEMP_C              (0x8) :   33.25 C     | (ok)
MB_DIMME_TEMP_C              (0x9) :   33.50 C     | (ok)
MB_DIMMF_TEMP_C              (0xA) :   33.00 C     | (ok)
MB_DIMMG_TEMP_C              (0xB) :   33.25 C     | (ok)
MB_DIMMH_TEMP_C              (0xC) :   33.00 C     | (ok)
MB_ADC_P12V_DIMM_VOLT_V      (0x25) :   12.36 Volts | (ok)
MB_VR_DIMMA_PMIC_PWR_W       (0x36) :    1.12 Watts | (ok)
MB_VR_DIMMB_PMIC_PWR_W       (0x37) :    0.62 Watts | (ok)
MB_VR_DIMMC_PMIC_PWR_W       (0x38) :    0.75 Watts | (ok)
MB_VR_DIMMD_PMIC_PWR_W       (0x39) :    0.50 Watts | (ok)
MB_VR_DIMME_PMIC_PWR_W       (0x3A) :    0.62 Watts | (ok)
MB_VR_DIMMF_PMIC_PWR_W       (0x3B) :    1.00 Watts | (ok)
MB_VR_DIMMG_PMIC_PWR_W       (0x3C) :    0.75 Watts | (ok)
MB_VR_DIMMH_PMIC_PWR_W       (0x3D) :    0.50 Watts | (ok)

2. Inject PMIC error and check SEL
- Inject DIMM warning error high_temp on DIMM A root@bmc-oob:~# log-util slot1 --print|grep -i ipmid
1    slot1    2023-08-15 14:02:15    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2023-08-15 14:02:14, Sensor: PMIC_ERR (0xB4), Event Data: (000C00) DIMM A High Temp Warning Assertion

### CODE REVIEW

**Review by williamspatrick — CHANGES_REQUESTED** [review]

File: common/hal/hal_i3c.c [hunk 1]
```diff
@@ -45,6 +46,7 @@ static struct i3c_dev_desc *find_matching_desc(const struct device *dev, uint8_t
 	for (i = 0; i < I3C_MAX_NUM; i++) {
 		desc = &i3c_desc_table[i];
 		if ((desc->master_dev == dev) && (desc->info.dynamic_addr == desc_addr)) {
+			*pos = i;
```
[the remark this answers is not in the document]
Comment (WayneLiu-wiwynn) [reply common/hal/hal_i3c.c:49]: OK.
I have followed the first option to revise.

[the remark this answers is not in the document]
Comment (williamspatrick) [reply common/hal/hal_i3c.c:49]: Don’t we need a null pointer check in this function then?

**Review by williamspatrick — CHANGES_REQUESTED** [review]

File: common/hal/hal_i3c.c [hunk 2]
```diff
@@ -143,6 +145,35 @@ int i3c_attach(I3C_MSG *msg)
...
+	int ret = 0, i, pos;
+	struct i3c_dev_desc *desc;
+
+	if (!dev_i3c[msg->bus]) {
+		LOG_ERR("Failed to detach address 0x%x due to undefined bus%u", msg->target_addr,
+			msg->bus);
+		return -ENODEV;
+	}
+
+	desc = find_matching_desc(dev_i3c[msg->bus], msg->target_addr, &pos);
```
Comment (jagpalgill) [inline common/hal/hal_i3c.c:161]: I suppose we need a NULL check for desc as well?

File: meta-facebook/yv35-gl/src/platform/plat_dimm.c [hunk 3]
```diff
@@ -9,18 +9,213 @@
...
+
+	// Init mutex
+	if (k_mutex_init(&i3c_dimm_mutex)) {
+		LOG_ERR("i3c_dimm_mux_mutex mutex init fail");
+	}
+
+	// Switch I3C mux to BIC when host post complete but BIC reset
+	if (get_post_status()) {
+		if (k_mutex_lock(&i3c_dimm_mutex, K_MSEC(I3C_DIMM_MUTEX_TIMEOUT_MS))) {
+			LOG_ERR("Failed to lock I3C dimm MUX");
```
Comment (jagpalgill) [inline meta-facebook/yv35-gl/src/platform/plat_dimm.c:74]: Do you still want to continue if lock failed?

File: meta-facebook/yv35-gl/src/platform/plat_dimm.c [hunk 4]
```diff
@@ -104,3 +307,161 @@ uint8_t sensor_num_map_dimm_id(uint8_t sensor_num)
...
+	// 0: switch I3C mux to CPU 1: switch I3C mux to BIC
+	i2c_msg.data[1] = i3c_mux_position;
+
+	ret = i2c_master_write(&i2c_msg, retry);
+	if (ret != 0) {
+		LOG_ERR("Failed to switch I3C MUX: 0x%x, ret= %d", i3c_mux_position, ret);
+		return false;
+	}
+
+	return false;
```
Comment (jagpalgill) [inline meta-facebook/yv35-gl/src/platform/plat_dimm.c:332]: You are retuning a false for a success case too i suppose, is that fine?

**Pull request closed without merging**